hoomd/hpmc/XenoSweep3D.h

```diff
@@ -0,0 +1,235 @@
+            {
+            // The origin is not within the projected minkowski sum on the plane
+            // perpendicular to the sweeping direction. No forward collision possible.
+            return resultNoCollision;
+            }
+
+        // Updating v1 or v2.
+        // consider the portals v1-v3 and v2-v3. The new portal should contain the
+        // origin on the far side of v0. If neither portal does, the origin is inside;
+        // then we have a portal for phase 2.
+        x  = cross( v3, v0 );
+        if( dot(x,v1) < OverlapReal(0.0) )
+            {
+            v2 = v3;
+            }
+        else if( dot(x,v2) > OverlapReal(0.0) )
+            {
+            v1 = v3;
+            }
+        else break;
+
+        // Finally update the normal vector for the next iteration.
+        n = cross(v0, v2-v1);
+        }
+
+
+    // Phase 2: Portal Refinement (in 3D)
+    count = 0;
+    while (true) // 3D-Refinement-Loop
+        {
+        count++;
+
+        n = cross(v1 - v2, v3 - v1);
+
+        // check if origin is inside (or overlapping, or behind, or touching)
+        // the = is important, because in an MC simulation you are guaranteed
+        // to find cases where edges and or vertices touch exactly.
+        if (dot(v1, n) <= OverlapReal(0.0))
+            {
+            collisionPlaneVector = n;
+            return resultNoForwardCollisionOrOverlapping - OverlapReal(count + 1);
+            }
+
+        // ----
+        // find support in direction of portal's outer facing normal
+        v4 = S(-n);
+
+        // Perform tolerance checks
+        // are we within an epsilon of the surface of the shape? If yes, done, one way
+        // or another
+        const OverlapReal tol_multiplier = 10000;
+        d = dot((v1 - v4) * tol_multiplier , n);
+        OverlapReal tol = precision_tol * tol_multiplier * R * fast::sqrt(dot(n,n));
+
+        // First, check if v4 is on plane (v2,v1,v3) or "behind"
+        if (d < tol)
+            {
+            collisionPlaneVector = n;
+            return dot(n,v1) / dot(n,v0);
+            }
+
+
+        if (count >= XENOCOLLIDE_3D_MAX_ITERATIONS)
+            {
+            err_count++;
+
+            collisionPlaneVector = n;
+            return dot(n,v1) / dot(n,v0);
+            // Return current distance, as it should be a better estimate than ignoring each other.
+            }
+
+        x = cross(v4, v0);
+        if (dot(v1, x) > OverlapReal(0.0))
+            {
+            if (dot(v2, x) > OverlapReal(0.0))
+                v1 = v4;    // Inside v1 & inside v2 ==> eliminate v1
+            else
+                v3 = v4;                   // Inside v1 & outside v2 ==> eliminate v3
+            }
+        else
+            {
+            if (dot(v3, x) > OverlapReal(0.0))
+                v2 = v4;    // Outside v1 & inside v3 ==> eliminate v2
+            else
+                v1 = v4;                   // Outside v1 & outside v3 ==> eliminate v1
+            }
+
```

Comment:
For readability I would set an exit at `if (count >= XENOCOLLIDE_3D_MAX_ITERATIONS)` and write that return statement outside the while loop. Even though you have to break it in the middle, it would be even nicer to use it as a condition for the while loop instead of the infinite loop over true (both for Phase 1 and 2).

Reply:
@b-butler If we did not finish yet, we need to update the portal. It is the same as in XenoCollide3D:
https://github.com/glotzerlab/hoomd-blue/blob/a5d516041247cfb81d81316dd0d9fa8c409115e3/hoomd/hpmc/XenoCollide3D.h#L250-L277

In each loop we get v4 as support vector to the normal of a triangle v1, v2, v3. If v4 is in the same plane as the triangle, we are at the surface, and can stop. Otherwise we need to continue. v4 with two of v1,v2,v3 form a triangle, that contains the origin in projection along the direction v0. The nested if statements pick that triangle. We need to iterate that triangle until it is at the surface of the Minkowski sum to calculate the distance.

Reply:
I missed the loop for some reason. I thought this was the end of the method.